lib/ex_doc/autolink.ex

```diff
@@ -0,0 +1,397 @@
+  #   checked if they are part of the app and based on that the links are relative or absolute.
+  #
+  # * `:current_module` - the module that the docs are being generated for. Used to link local
+  #   calls and see if remote calls are in the same app.
+  #
+  # * `:module_id` - id of the module being documented (e.g.: `"String"`)
+  #
+  # * `:id` - id of the thing being documented (e.g.: `"String.upcase/2"`, `"library-guidelines"`, etc)
+  #
+  # * `:ext` - the extension (`".html"`, "`.xhtml"`, etc)
+  #
+  # * `:skip_undefined_reference_warnings_on` - list of modules to skip the warning on
+  defstruct [
+    :app,
+    :current_module,
+    :module_id,
+    :id,
+    ext: ".html",
+    skip_undefined_reference_warnings_on: []
+  ]
+
+  alias ExDoc.Formatter.HTML.Templates, as: T
+  alias ExDoc.Refs
+
+  @hexdocs "https://hexdocs.pm/"
+  @otpdocs "http://www.erlang.org/doc/man/"
+
+  def doc(ast, options \\ []) do
+    config = struct!(__MODULE__, options)
+    walk(ast, config)
+  end
+
+  defp walk(list, config) when is_list(list) do
+    Enum.map(list, &walk(&1, config))
+  end
+
+  defp walk(binary, _) when is_binary(binary) do
+    binary
+  end
+
+  defp walk({:pre, _, _} = ast, _config) do
+    ast
+  end
+
+  defp walk({:a, attrs, _} = ast, config) do
+    if url = custom_link(attrs, config) do
+      {:a, Keyword.put(attrs, :href, url), ast}
+    else
+      ast
+    end
+  end
+
+  defp walk({:code, _, [text]} = ast, config) do
+    case text_to_ref(text, :regular, config) do
+      :no_ref ->
+        ast
+
+      ref ->
+        if url = url(ref, config) do
+          {:a, [href: url], [ast]}
+        else
+          ast
+        end
+    end
+  end
+
+  defp walk({tag, attrs, ast}, config) do
+    {tag, attrs, walk(ast, config)}
+  end
+
+  defp custom_link(attrs, config) do
+    with {:ok, href} <- Keyword.fetch(attrs, :href),
+         [[_, text]] <- Regex.scan(~r/^`(.+)`$/, href) do
+      case text_to_ref(text, :custom_link, config) do
+        :no_ref ->
+          nil
+
+        ref ->
+          url(ref, config)
+      end
+    else
+      _ -> nil
+    end
+  end
+
+  defp text_to_ref("c:" <> text, mode, config) do
+    with {call, {:function, module, name, arity}} <- text_to_ref(text, mode, config) do
+      {call, {:callback, module, name, arity}}
+    end
+  end
+
+  defp text_to_ref("t:" <> text, mode, config) do
+    with {call, {:function, module, name, arity}} <- text_to_ref(text, mode, config) do
+      {call, {:type, module, name, arity}}
+    end
+  end
+
+  defp text_to_ref("mix help " <> name, _mode, _config), do: mix_task(name)
+  defp text_to_ref("mix " <> name, _mode, _config), do: mix_task(name)
+
+  defp text_to_ref(text, mode, config) do
+    if not String.contains?(text, [" ", "(", ")"]) do
+      case Code.string_to_quoted(text) do
+        {:ok, {:__aliases__, _, _} = module} ->
+          {:module, module(module)}
```

Comment:
In my ideal world, this would already call ExDoc.Refs.module(...) and handle it accordingly. Any reason why to have an intermediate step?

Reply:
good call!